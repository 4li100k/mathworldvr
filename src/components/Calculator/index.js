import React from 'react'
import PropTypes from 'prop-types'
import { Entity } from 'aframe-react'
import { CalcButton, Text } from 'components'

const propTypes = {
  displayText: PropTypes.string.isRequired,
  writeText: PropTypes.func.isRequired,
  backspace: PropTypes.func.isRequired,
  clearText: PropTypes.func.isRequired,
  updateEquation: PropTypes.func.isRequired,
}

const Calculator = ({ displayText, writeText, backspace, clearText, updateEquation }) => {
  return (
    <Entity
      className="interactive calculator"
      id="text-box"
      geometry="primitive: box; width: 0.88; height: 0.65; depth: 0.01;"
      material="shader: flat; side: double; color: #8d8547; "
      position="-0.03 1.2 -0.26"
      rotation="-45 0 0"
      scale="1 1 1"
      grabbable
      stretchable
      dynamic-body
    >
      { /* Calculator display */ }
      <Text
        id="calc-display"
        value={displayText}
        align="center"
        color="#fff"
        position={{ x: 0, y: 0.26, z: 0 }}
        width={0.5}
        zOffset={0.01}
      />

      { /* --- HERE STARTS THE BUTTONS --- */ }
      <Entity position={{ x: 0, y: 0.08, z: 0 }}>
        { /* Update button */ }
        <CalcButton
          id="calc-update"
          value="Update"
          position={{ x: -0.28, y: -0.19, z: 0.03 }}
          width={0.43}
          actionToTrigger={() => updateEquation(displayText)}
        />

        <CalcButton value="sin" position={{ x: -0.37, y: 0.09, z: 0.03 }} id="calc-sin" actionToTrigger={() => writeText('sin(')} />
        <CalcButton value="cos" position={{ x: -0.28, y: 0.09, z: 0.03 }} id="calc-cos" actionToTrigger={() => writeText('cos(')} />
        <CalcButton value="tan" position={{ x: -0.19, y: 0.09, z: 0.03 }} id="calc-tan" actionToTrigger={() => writeText('tan(')} />
        <CalcButton value="sqrt" position={{ x: -0.37, y: 0, z: 0.03 }} id="calc-sqrt" actionToTrigger={() => writeText('sqrt(')} />
        <CalcButton value="log" position={{ x: -0.28, y: 0, z: 0.03 }} id="calc-log" actionToTrigger={() => writeText('log(')} />
        <CalcButton value="ln" position={{ x: -0.19, y: 0, z: 0.03 }} id="calc-ln" actionToTrigger={() => writeText('ln(')} />
        <CalcButton value="e" position={{ x: -0.37, y: -0.09, z: 0.03 }} id="calc-e" actionToTrigger={() => writeText('e')} />
        <CalcButton value="abs" position={{ x: -0.28, y: -0.09, z: 0.03 }} id="calc-abs" actionToTrigger={() => writeText('abs(')} />
        <CalcButton value="^" position={{ x: -0.19, y: -0.09, z: 0.03 }} id="calc-square" actionToTrigger={() => writeText('^')} />

        <CalcButton value="(" position={{ x: 0.19, y: 0.09, z: 0.03 }} id="calc-parent-left" actionToTrigger={() => writeText('(')} />
        <CalcButton value=")" position={{ x: 0.19, y: 0, z: 0.03 }} id="calc-parent-right" actionToTrigger={() => writeText(')')} />
        <CalcButton value="+" position={{ x: 0.28, y: 0.09, z: 0.03 }} id="calc-plus" actionToTrigger={() => writeText('+')} />
        <CalcButton value="-" position={{ x: 0.28, y: 0, z: 0.03 }} id="calc-minus" actionToTrigger={() => writeText('-')} />
        <CalcButton value="*" position={{ x: 0.28, y: -0.09, z: 0.03 }} id="calc-multiply" actionToTrigger={() => writeText('*')} />
        <CalcButton value="/" position={{ x: 0.28, y: -0.18, z: 0.03 }} id="calc-divide" actionToTrigger={() => writeText('/')} />
        <CalcButton value="exp" position={{ x: 0.19, y: -0.09, z: 0.03 }} id="calc-exp" actionToTrigger={() => writeText('exp(')} />
        <CalcButton value="pi" position={{ x: 0.19, y: -0.18, z: 0.03 }} id="calc-pi" actionToTrigger={() => writeText('pi')} />

        <CalcButton value="1" position={{ x: -0.09, y: -0.09, z: 0.03 }} id="calc-1" actionToTrigger={() => writeText('1')} />
        <CalcButton value="2" position={{ x: 0, y: -0.09, z: 0.03 }} id="calc-2" actionToTrigger={() => writeText('2')} />
        <CalcButton value="3" position={{ x: 0.09, y: -0.09, z: 0.03 }} id="calc-3" actionToTrigger={() => writeText('3')} />
        <CalcButton value="4" position={{ x: -0.09, y: 0, z: 0.03 }} id="calc-4" actionToTrigger={() => writeText('4')} />
        <CalcButton value="5" position={{ x: 0, y: 0, z: 0.03 }} id="calc-5" actionToTrigger={() => writeText('5')} />
        <CalcButton value="6" position={{ x: 0.09, y: 0, z: 0.03 }} id="calc-6" actionToTrigger={() => writeText('6')} />
        <CalcButton value="7" position={{ x: -0.09, y: 0.09, z: 0.03 }} id="calc-7" actionToTrigger={() => writeText('7')} />
        <CalcButton value="8" position={{ x: 0, y: 0.09, z: 0.03 }} id="calc-8" actionToTrigger={() => writeText('8')} />
        <CalcButton value="9" position={{ x: 0.09, y: 0.09, z: 0.03 }} id="calc-9" actionToTrigger={() => writeText('9')} />
        <CalcButton value="0" position={{ x: 0, y: -0.18, z: 0.03 }} id="calc-0" actionToTrigger={() => writeText('0')} />
        <CalcButton value="." position={{ x: -0.09, y: -0.18, z: 0.03 }} id="calc-dot" actionToTrigger={() => writeText('.')} />
        <CalcButton value="," position={{ x: 0.09, y: -0.18, z: 0.03 }} id="calc-comma" actionToTrigger={() => writeText(',')} />

        <CalcButton value="<-" position={{ x: 0.37, y: 0.09, z: 0.03 }} id="calc-backspace" actionToTrigger={backspace} />
        <CalcButton value="clr" position={{ x: 0.37, y: 0, z: 0.03 }} id="calc-clear" actionToTrigger={clearText} />
        <CalcButton value="x" position={{ x: 0.37, y: -0.09, z: 0.03 }} id="calc-x" actionToTrigger={() => writeText('x')} />
        <CalcButton value="y" position={{ x: 0.37, y: -0.18, z: 0.03 }} id="calc-y" actionToTrigger={() => writeText('y')} />
      </Entity>
      { /* --- END OF BUTTONS --- */ }

      <Entity id="sld-disclaimer" position={{ x: -0.09, y: -0.29, z: 0.02 }}>
        <Entity
          id="sld-logo"
          geometry={{ primitive: 'plane', width: 1, height: 1 }}
          material={{ transparent: true, opacity: 1, shader: 'flat', src: 'url(sld-logo.png)', side: 'front', height: 1024, width: 1024 }}
          position={{ x: -0.18, y: 0.05, z: 0.01 }}
          scale={{ x: 0.14, y: 0.14, z: 0.14 }}
        />
        <Text
          color="#fff"
          align="center"
          value="Disclaimer: This is an experimental project\n
            made by VR team at Sleighdogs GmbH with some help from Technical University in Kosice (Slovakia).\n
            If grabbing with HTC Vive / Oculus Touch controllers doesn't work,\n
            try to reload the page with your controllers connected!\n
            http://sld.gs/"
          position={{ x: 0.24, y: 0.05, z: -0.01 }}
          width={0.3}
          lineHeight={50}
        />
      </Entity>
    </Entity>
  )
}

Calculator.propTypes = propTypes
export default Calculator
