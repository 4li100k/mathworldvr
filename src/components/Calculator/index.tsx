import React from "react";
import { Entity } from "aframe-react";
import { CalcButton, Text } from "../";

export type CalculatorProps = {
  displayText: string;
  writeText: () => Promise<void>;
  backspace: () => Promise<void>;
  clearText: () => Promise<void>;
  updateEquation: () => Promise<void>;
};

export function Calculator({
  displayText,
  writeText,
  backspace,
  clearText,
  updateEquation
}: CalculatorProps) {
  return (
    <Entity
      className="interactive calculator"
      id="text-box"
      geometry="primitive: box; width: 0.88; height: 0.65; depth: 0.01;"
      material="shader: flat; side: double; color: #8d8547; "
      position="-0.03 1.2 -0.26"
      rotation="-45 0 0"
      scale="1 1 1"
      grabbable
      stretchable
      dynamic-body
    >
      {/* Calculator display */}
      <Text
        id="calc-display"
        value={displayText}
        align="center"
        color="#fff"
        position={{ x: 0, y: 0.26, z: 0 }}
        width={0.5}
        zOffset={0.01}
      />

      {/* --- HERE STARTS THE BUTTONS --- */}
      <Entity position={{ x: 0, y: 0.08, z: 0 }}>
        {/* Update button */}
        <CalcButton
          id="calc-update"
          text="Update"
          value={displayText}
          position={{ x: -0.28, y: -0.19, z: 0.03 }}
          width={0.43}
          actionToTrigger={updateEquation}
        />

        <CalcButton
          text="sin"
          position={{ x: -0.37, y: 0.09, z: 0.03 }}
          id="calc-sin"
          value="sin("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="cos"
          position={{ x: -0.28, y: 0.09, z: 0.03 }}
          id="calc-cos"
          value="cos("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="tan"
          position={{ x: -0.19, y: 0.09, z: 0.03 }}
          id="calc-tan"
          value="tan("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="sqrt"
          position={{ x: -0.37, y: 0, z: 0.03 }}
          id="calc-sqrt"
          value="sqrt("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="log"
          position={{ x: -0.28, y: 0, z: 0.03 }}
          id="calc-log"
          value="log("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="ln"
          position={{ x: -0.19, y: 0, z: 0.03 }}
          id="calc-ln"
          value="ln("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="e"
          position={{ x: -0.37, y: -0.09, z: 0.03 }}
          id="calc-e"
          value="e"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="abs"
          position={{ x: -0.28, y: -0.09, z: 0.03 }}
          id="calc-abs"
          value="abs("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="^"
          position={{ x: -0.19, y: -0.09, z: 0.03 }}
          id="calc-square"
          value="^"
          actionToTrigger={writeText}
        />

        <CalcButton
          text="("
          position={{ x: 0.19, y: 0.09, z: 0.03 }}
          id="calc-parent-left"
          value="("
          actionToTrigger={writeText}
        />
        <CalcButton
          text=")"
          position={{ x: 0.19, y: 0, z: 0.03 }}
          id="calc-parent-right"
          value=")"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="+"
          position={{ x: 0.28, y: 0.09, z: 0.03 }}
          id="calc-plus"
          value="+"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="-"
          position={{ x: 0.28, y: 0, z: 0.03 }}
          id="calc-minus"
          value="-"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="*"
          position={{ x: 0.28, y: -0.09, z: 0.03 }}
          id="calc-multiply"
          value="*"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="/"
          position={{ x: 0.28, y: -0.18, z: 0.03 }}
          id="calc-divide"
          value="/"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="exp"
          position={{ x: 0.19, y: -0.09, z: 0.03 }}
          id="calc-exp"
          value="exp("
          actionToTrigger={writeText}
        />
        <CalcButton
          text="pi"
          position={{ x: 0.19, y: -0.18, z: 0.03 }}
          id="calc-pi"
          value="pi"
          actionToTrigger={writeText}
        />

        <CalcButton
          text="1"
          position={{ x: -0.09, y: -0.09, z: 0.03 }}
          id="calc-1"
          value="1"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="2"
          position={{ x: 0, y: -0.09, z: 0.03 }}
          id="calc-2"
          value="2"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="3"
          position={{ x: 0.09, y: -0.09, z: 0.03 }}
          id="calc-3"
          value="3"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="4"
          position={{ x: -0.09, y: 0, z: 0.03 }}
          id="calc-4"
          value="4"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="5"
          position={{ x: 0, y: 0, z: 0.03 }}
          id="calc-5"
          value="5"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="6"
          position={{ x: 0.09, y: 0, z: 0.03 }}
          id="calc-6"
          value="6"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="7"
          position={{ x: -0.09, y: 0.09, z: 0.03 }}
          id="calc-7"
          value="7"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="8"
          position={{ x: 0, y: 0.09, z: 0.03 }}
          id="calc-8"
          value="8"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="9"
          position={{ x: 0.09, y: 0.09, z: 0.03 }}
          id="calc-9"
          value="9"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="0"
          position={{ x: 0, y: -0.18, z: 0.03 }}
          id="calc-0"
          value="0"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="."
          position={{ x: -0.09, y: -0.18, z: 0.03 }}
          id="calc-dot"
          value="."
          actionToTrigger={writeText}
        />
        <CalcButton
          text=","
          position={{ x: 0.09, y: -0.18, z: 0.03 }}
          id="calc-comma"
          value=","
          actionToTrigger={writeText}
        />

        <CalcButton
          text="<-"
          position={{ x: 0.37, y: 0.09, z: 0.03 }}
          id="calc-backspace"
          actionToTrigger={backspace}
        />
        <CalcButton
          text="clr"
          position={{ x: 0.37, y: 0, z: 0.03 }}
          id="calc-clear"
          actionToTrigger={clearText}
        />
        <CalcButton
          text="x"
          position={{ x: 0.37, y: -0.09, z: 0.03 }}
          id="calc-x"
          value="x"
          actionToTrigger={writeText}
        />
        <CalcButton
          text="y"
          position={{ x: 0.37, y: -0.18, z: 0.03 }}
          id="calc-y"
          value="y"
          actionToTrigger={writeText}
        />
      </Entity>
      {/* --- END OF BUTTONS --- */}
    </Entity>
  );
}
