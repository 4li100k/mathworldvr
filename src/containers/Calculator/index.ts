// TODO: refactor to support state management with React.useContext and constate

// import { connect } from 'react-redux'
// import { Calculator } from 'components'
// import { calculatorWriteText, calculatorBackspace, calculatorClearText, parametricFunctionSetEquation } from 'actions'

// const mapStateToProps = (state) => ({
//   displayText: state.calculator.displayText,
// })

// const mapDispatchToProps = (dispatch) => ({
//   writeText: (text) => dispatch(calculatorWriteText(text)),
//   backspace: () => dispatch(calculatorBackspace()),
//   clearText: () => dispatch(calculatorClearText()),
//   updateEquation: (equation) => dispatch(parametricFunctionSetEquation(equation)),
// })

// export default connect(mapStateToProps, mapDispatchToProps)(Calculator)

import * as React from "react";
import createContainer from "constate";

export const CALCULATOR_INITIAL_STATE = {
  expression: "x^2 + y^2",
  variables: ["x", "y"],
  inputs: []
};

function useCalculator({ initialState = CALCULATOR_INITIAL_STATE } = {}) {
  const [settings, updateState] = React.useState(initialState);
  const updateEquation = (expression: string) =>
    updateState({
      ...settings,
      expression
    });
  const setVariables = (variables: string[]) =>
    updateState({
      ...settings,
      variables
    });
  const setInputs = (inputs: string[]) =>
    updateState({
      ...settings,
      inputs
    });
  return { settings, updateEquation, setVariables, setInputs };
}

export const CalculatorContainer = createContainer(useCalculator, settings => [
  settings
]);
