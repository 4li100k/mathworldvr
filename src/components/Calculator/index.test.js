import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Calculator from '.'

jest.mock('react-dom')

describe('Calculator', () => {
  it('renders <a-entity>', () => {
    const tree = renderer.create(<Calculator />).toJSON()
    expect(tree.type).toBe('a-entity')
  })

  it('renders the entity consistently', () => {
    const entityWrapper = shallow(<Calculator />)
    expect(entityWrapper).toMatchSnapshot()
  })
})
