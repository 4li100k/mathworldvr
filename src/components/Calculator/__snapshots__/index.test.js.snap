// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders the entity consistently 1`] = `
<a-entity
  className="interactive calculator"
  id="text-box"
>
  <a-entity />
  <a-entity>
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
    <a-entity />
  </a-entity>
</a-entity>
`;
